src/app/clusters/thermostat-server/thermostat-server.cpp

```diff
@@ -1287,102 +1306,122 @@ bool emberAfThermostatClusterSetActivePresetRequestCallback(
 
-    // If presets are not editable, return INVALID_IN_STATE.
-    if (!gThermostatAttrAccess.GetPresetsEditable(endpoint))
+    if (!commandData.timeout.HasValue())
     {
-        commandObj->AddStatus(commandPath, imcode::InvalidInState);
+        commandObj->AddStatus(commandPath, imcode::InvalidCommand);
         return true;
     }
 
-    ScopedNodeId sourceNodeId = GetSourceScopedNodeId(commandObj);
+    auto timeout = commandData.timeout.Value();
 
-    // If the node id sending the CancelPresetsSchedulesRequest command is not the same as the one which send the
-    // previous StartPresetsSchedulesEditRequest, return UNSUPPORTED_ACCESS.
-    if (gThermostatAttrAccess.GetOriginatorScopedNodeId(endpoint) != sourceNodeId)
+    if (!validAtomicAttributes(commandData))
     {
-        commandObj->AddStatus(commandPath, imcode::UnsupportedAccess);
+        commandObj->AddStatus(commandPath, imcode::InvalidCommand);
         return true;
     }
 
-    Delegate * delegate = GetDelegate(endpoint);
-
-    if (delegate == nullptr)
+    bool inAtomicWrite = gThermostatAttrAccess.InAtomicWrite(endpoint);
+    if (inAtomicWrite)
     {
-        ChipLogError(Zcl, "Delegate is null");
-        return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::InvalidInState);
+        chip::app::Clusters::Thermostat::Commands::AtomicResponse::Type response;
+        Globals::Structs::AtomicAttributeStatusStruct::Type attributeStatus[] = {
+            { .attributeID = Presets::Id, .statusCode = static_cast<uint8_t>(imcode::Busy) },
+            { .attributeID = Schedules::Id, .statusCode = static_cast<uint8_t>(imcode::Busy) }
+        };
+
+        response.statusCode      = static_cast<uint8_t>(imcode::Failure);
+        response.attributeStatus = attributeStatus;
+        commandObj->AddResponse(commandPath, response);
+        // commandObj->AddStatus(commandPath, imcode::Busy);
+        return true;
     }
 
-    // Clear the timer, discard the changes and set PresetsEditable to false.
-    return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::Success);
+    uint16_t maxTimeout = 5000;
+    timeout             = std::min(timeout, maxTimeout);
+
+    ScheduleTimer(endpoint, timeout);
+    gThermostatAttrAccess.SetAtomicWrite(endpoint, true);
+    gThermostatAttrAccess.SetAtomicWriteScopedNodeId(endpoint, GetSourceScopedNodeId(commandObj));
+    chip::app::Clusters::Thermostat::Commands::AtomicResponse::Type response;
+    Globals::Structs::AtomicAttributeStatusStruct::Type attributeStatus[] = {
+        { .attributeID = Presets::Id, .statusCode = static_cast<uint8_t>(imcode::Success) },
+        { .attributeID = Schedules::Id, .statusCode = static_cast<uint8_t>(imcode::Success) }
+    };
+
+    response.statusCode      = static_cast<uint8_t>(imcode::Failure);
+    response.attributeStatus = attributeStatus;
+    response.timeout.Emplace(timeout);
+    commandObj->AddResponse(commandPath, response);
```

Comment:
Things that seem to be missing here, compared to the spec: privilege checks.  Presumably that's because the fact that we managed to invoke the command at all means we have sufficient privileges to write to the attributes, but we should have a clear comment saying that, so that when someone copies/moves this code for use in another context we notice that the assumption might not hold anymore.